ENH: added zoomed in view of the slice under the cursor.

ayarmarkovich opened the pull request:
This functionality was available in Slicer2 and has been requested by Raul's COPD group. It creates a zoomed-in view of a slice centered on the mouse cursor in the Data Probe panel. The zoom factor is currently set to 10. Should it be configurable?

pieper commented:
A few things:

* The check box takes up extra space in the DataProbe -- consensus on the mailing list was that we the DataProbe already takes up too much space.  Can you shrink it?  For example the checkbox doesn't need to show when the data probe is displaying info so maybe it can share space with the numerical readout.

* I played around with it and got a crash after just a few moments.  I'll paste the top part of the crash below.  Some of the console messages were:

QPixmap::scaled: Pixmap is a null pixmap
Input port 0 of algorithm vtkImageBlend(0x7f993d1a0210) has 0 connections but is not optional.

(repeated many times)

The update extent specified in the information for output port 0 on algorithm vtkImageBlend(0x7f993d1a0210) is -10 22 6 38 0 0, which is outside the whole extent 0 525 0 314 0 0.


The update extent specified in the information for output port 0 on algorithm vtkImageBlend(0x7f993d1a0210) is -3 29 174 206 0 0, which is outside the whole extent 0 525 0 314 0 0.


The update extent specified in the information for output port 0 on algorithm vtkImageBlend(0x7f993d1a0210) is -3 29 174 206 0 0, which is outside the whole extent 0 525 0 314 0 0.

(Then finally this crash)

Thread 0 Crashed:: Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	0x00007fff8a06e286 __pthread_kill + 10
1   libsystem_c.dylib             	0x00007fff8d39ab53 abort + 129
2   libsystem_c.dylib             	0x00007fff8d362c39 __assert_rtn + 321
3   libvtkCommon-6.2.1.dylib      	0x000000011e43232b vtkExtractStructuredGridHelper::CopyPointsAndPointData(int*, int*, vtkPointData*, vtkPoints*, vtkPointData*, vtkPoints*, int*) + 1307 (vtkExtractStructuredGridHelper.cxx:376)
4   libvtkImaging-6.2.1.dylib     	0x000000011d797a08 vtkExtractVOI::RequestDataImpl(int*, vtkInformationVector**, vtkInformationVector*) + 3128 (vtkExtractVOI.cxx:274)
5   libvtkImaging-6.2.1.dylib     	0x000000011d796da9 vtkExtractVOI::RequestData(vtkInformation*, vtkInformationVector**, vtkInformationVector*) + 73 (vtkExtractVOI.cxx:175)
6   libvtkCommon-6.2.1.dylib      	0x000000011e79552a vtkImageAlgorithm::ProcessRequest(vtkInformation*, vtkInformationVector**, vtkInformationVector*) + 122 (vtkImageAlgorithm.cxx:103)
7   libvtkCommon-6.2.1.dylib      	0x000000011e77f154 vtkExecutive::CallAlgorithm(vtkInformation*, int, vtkInformationVector**, vtkInformationVector*) + 164 (vtkExecutive.cxx:775)
8   libvtkCommon-6.2.1.dylib      	0x000000011e7741fa vtkDemandDrivenPipeline::ExecuteData(vtkInformation*, vtkInformationVector**, vtkInformationVector*) + 154 (vtkDemandDrivenPipeline.cxx:491)
9   libvtkCommon-6.2.1.dylib      	0x000000011e76b8e5 vtkCompositeDataPipeline::ExecuteData(vtkInformation*, vtkInformationVector**, vtkInformationVector*) + 1797 (vtkCompositeDataPipeline.cxx:178)
10  libvtkCommon-6.2.1.dylib      	0x000000011e77352d vtkDemandDrivenPipeline::ProcessRequest(vtkInformation*, vtkInformationVector**, vtkInformationVector*) + 1357 (vtkDemandDrivenPipeline.cxx:274)
11  libvtkCommon-6.2.1.dylib      	0x000000011e7ad67b vtkStreamingDemandDrivenPipeline::ProcessRequest(vtkInformation*, vtkInformationVector**, vtkInformationVector*) + 2843 (vtkStreamingDemandDrivenPipeline.cxx:310)
12  libvtkCommon-6.2.1.dylib      	0x000000011e773fb8 vtkDemandDrivenPipeline::UpdateData(int) + 1096 (vtkDemandDrivenPipeline.cxx:441)
13  libvtkCommon-6.2.1.dylib      	0x000000011e7ad99e vtkStreamingDemandDrivenPipeline::Update(int) + 270 (vtkStreamingDemandDrivenPipeline.cxx:363)
14  libvtkCommon-6.2.1.dylib      	0x000000011e7613f2 vtkAlgorithm::Update(int) + 50 (vtkAlgorithm.cxx:1456)
15  libvtkCommon-6.2.1.dylib      	0x000000011e7613b8 vtkAlgorithm::Update() + 72 (vtkAlgorithm.cxx:1450)
16  libvtkCommonKitPython27D-6.2.1.dylib	0x000000012ac75cfb PyvtkAlgorithm_Update_s2(_object*, _object*) + 155 (vtkAlgorithmPython.cxx:2250)
17  libvtkCommonKitPython27D-6.2.1.dylib	0x000000012ac73125 PyvtkAlgorithm_Update(_object*, _object*) + 101 (vtkAlgorithmPython.cxx:2275)
18  libpython2.7.dylib            	0x000000011ef87d11 PyEval_EvalFrameEx + 19617
19  libpython2.7.dylib            	0x000000011ef82d66 PyEval_EvalCodeEx + 1542
20  libpython2.7.dylib            	0x000000011ef2186c function_call + 364
21  libpython2.7.dylib            	0x000000011eef8793 PyObject_Call + 99
22  libpython2.7.dylib            	0x000000011ef05826 instancemethod_call + 182
23  libpython2.7.dylib            	0x000000011eef8793 PyObject_Call + 99
24  libpython2.7.dylib            	0x000000011ef89fbd PyEval_CallObjectWithKeywords + 93
25  libvtkWrappingPython27Core-6.2.1.dylib	0x0000000117a90ed9 vtkPythonCommand::Execute(vtkObject*, unsigned long, void*) + 1241 (vtkPythonCommand.cxx:222)
26  libvtkCommon-6.2.1.dylib      	0x000000011e1e575b vtkSubjectHelper::InvokeEvent(unsigned long, void*, vtkObject*) + 4731 (vtkObject.cxx:616)
27  libvtkCommon-6.2.1.dylib      	0x000000011e1e5d32 vtkObject::InvokeEvent(unsigned long, void*) + 66 (vtkObject.cxx:785)
28  libMRMLCore.dylib             	0x000000010df439c2 vtkMRMLCrosshairNode::SetCursorPositionXYZ(double*, vtkMRMLSliceNode*) + 370 (vtkMRMLCrosshairNode.cxx:317)
29  libMRMLDisplayableManager.dylib	0x000000010acdcb28 vtkMRMLCrosshairDisplayableManager::OnInteractorStyleEvent(int) + 680 (vtkMRMLCrosshairDisplayableManager.cxx:721)
30  libMRMLDisplayableManager.dylib	0x000000010ad02c24 vtkMRMLAbstractDisplayableManager::vtkInternal::DoInteractorStyleCallback(vtkObject*, unsigned long, void*, void*) + 212 (vtkMRMLAbstractDisplayableManager.cxx:372)
31  libvtkCommon-6.2.1.dylib      	0x000000011dfc9374 vtkCallbackCommand::Execute(vtkObject*, unsigned long, void*) + 84 (vtkCallbackCommand.cxx:43)
32  libvtkCommon-6.2.1.dylib      	0x000000011e1e575b vtkSubjectHelper::InvokeEvent(unsigned long, void*, vtkObject*) + 4731 (vtkObject.cxx:616)
33  libvtkCommon-6.2.1.dylib      	0x000000011e1e5d32 vtkObject::InvokeEvent(unsigned long, void*) + 66 (vtkObject.cxx:785)
34  libvtkInteraction-6.2.1.dylib 	0x0000000119da41e9 vtkInteractorStyleUser::OnEnter() + 169 (vtkInteractorStyleUser.cxx:334)
35  libMRMLDisplayableManager.dylib	0x000000010ad1cd15 vtkSliceViewInteractorStyle::OnEnter() + 21 (vtkSliceViewInteractorStyle.cxx:419)
36  libvtkRendering-6.2.1.dylib   	0x000000011b658c6e vtkInteractorStyle::ProcessEvents(vtkObject*, unsigned long, void*, void*) + 478 (vtkInteractorStyle.cxx:994)
37  libvtkCommon-6.2.1.dylib      	0x000000011dfc9374 vtkCallbackCommand::Execute(vtkObject*, unsigned long, void*) + 84 (vtkCallbackCommand.cxx:43)
38  libvtkCommon-6.2.1.dylib      	0x000000011e1e575b vtkSubjectHelper::InvokeEvent(unsigned long, void*, vtkObject*) + 4731 (vtkObject.cxx:616)
39  libvtkCommon-6.2.1.dylib      	0x000000011e1e5d32 vtkObject::InvokeEvent(unsigned long, void*) + 66 (vtkObject.cxx:785)
40  libvtkGUISupportQt-6.2.1.dylib	0x0000000117cd8e8d QVTKInteractorAdapter::ProcessEvent(QEvent*, vtkRenderWindowInteractor*) + 1357 (QVTKInteractorAdapter.cxx:165)
41  libvtkGUISupportQt-6.2.1.dylib	0x0000000117cdc13a QVTKWidget::enterEvent(QEvent*) + 90 (QVTKWidget.cxx:535)
42  QtGui                         	0x0000000117e42311 QWidget::event(QEvent*) + 2177
43  libvtkGUISupportQt-6.2.1.dylib	0x0000000117cdbaad QVTKWidget::event(QEvent*) + 413 (QVTKWidget.cxx:412)
44  QtGui                         	0x0000000117deab6d QApplicationPrivate::notify_helper(QObject*, QEvent*) + 189
45  QtGui                         	0x0000000117deedc8 QApplication::notify(QObject*, QEvent*) + 232
46  libqSlicerBaseQTGUI.dylib     	0x0000000107caed3f qSlicerApplication::notify(QObject*, QEvent*) + 63 (qSlicerApplication.cxx:338)
47  QtCore                        	0x0000000118d3868c QCoreApplication::notifyInternal(QObject*, QEvent*) + 124
48  QtGui                         	0x0000000117df1fde QApplicationPrivate::dispatchEnterLeave(QWidget*, QWidget*) + 1774
49  QtGui                         	0x0000000117d9e97b qt_mac_handleMouseEvent(NSEvent*, QEvent::Type, Qt::MouseButton, QWidget*, bool) + 1515
50  com.apple.AppKit              	0x00007fff888d5fad -[NSTrackingArea _dispatchMouseEntered:] + 162
51  com.apple.AppKit              	0x00007fff8883a897 -[NSApplication sendEvent:] + 3917
52  QtGui                         	0x0000000117d987d4 -[QNSApplication sendEvent:] + 84
53  com.apple.AppKit              	0x00007fff88763c68 -[NSApplication run] + 711
54  QtGui                         	0x0000000117da3640 QEventDispatcherMac::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) + 1824
55  QtCore                        	0x0000000118d375a4 QEventLoop::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) + 68
56  QtCore                        	0x0000000118d37954 QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) + 324
57  QtCore                        	0x0000000118d3a1cc QCoreApplication::exec() + 188
58                                	0x00000001078e579b (anonymous namespace)::SlicerAppMain(int, char**) + 5707
59                                	0x00000001078e4142 main + 34
60  libdyld.dylib                 	0x00007fff96c1e5c9 start + 1

pieper commented:
Crash seems to happen when moving between slice views.

jcfr commented:
Instead of implementing the logic code in the DataProbe, I would highly recommend to use `ctkVTKMagnifyView`. It has originally been designed for that specific purpose.

See https://github.com/commontk/CTK/blob/master/Libs/Visualization/VTK/Widgets/ctkVTKMagnifyView.h

It is also interesting to read the comment explaining why it has been removed from Slice4. See [r18030](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=18030)

pieper commented:
I'm not sure the ctk code actually simplifies things.  It adds more layers
of code and it has proven hard for developers sometimes to find their way
to the code that implements a particular feature (this has certainly
happened to me, but I have heard similar reports from other people).

If you look at the current implementation from Alex there really isn't much
code and it's pretty clear.  So unless we are adding a lot more features I
don't have an objection except for the issues I've noted so far.

I'd have to go back and look but as it says in the commit message the
version that was originally implemented in ctk and included slicer4 was not
serving the required purpose so additional debugging would be required
there.

On Tue, May 12, 2015 at 5:22 PM, Jean-Christophe Fillion-Robin <
notifications@github.com> wrote:

> Instead of implementing the logic code in the DataProbe, I would highly
> recommend to use ctkVTKMagnifyView. It has originally been designed for
> that specific purpose.
>
> See
> https://github.com/commontk/CTK/blob/master/Libs/Visualization/VTK/Widgets/ctkVTKMagnifyView.h
>
> It is also interesting to read the comment explaining why it has been
> removed from Slice4. See r18030
> <http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=18030>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Slicer/Slicer/pull/276#issuecomment-101425833>.
>

ayarmarkovich commented:
Hi Steve, I am not sure I completely understand you UI suggestion, do you want it to show numerical values when you mouse in slicer view and the checkbox when your mouse is in the panel? It may be confusing. I can move the checkbox to the right of the numerical values to save vertical space if this sounds ok to you.

Also I cannot reproduce the crash you reported on windows. Are you using Mac? I will try to add more checks in the code and perhaps you can try again.

pieper commented:
Hi Alex -

I don't think we should make the data probe wider either.

I'm suggesting that :

1) when the data probe is showing numbers it is because the mouse is in a
slice view and in that case you cannot check the box so the box could go
away.

2) when the mouse is not in the data probe then it is impossible to check
the checkbox so it does not have to be seen.

3) when the mouse is not in a slice view the data probe has no numbers, so
the checkbox could go where the numbers were previously.

Essentially this would go back to something like the original slicer2
configuration, we could use the space in the dataprobe to show the slicer
logo when the mouse is outside the the slice view.  then when you mouse
over the slicer logo you see a panel to configure the data probe.

Yes, the crash occurs on the mac.

-Steve

On Tue, May 12, 2015 at 5:51 PM, ayarmarkovich <notifications@github.com>
wrote:

> Hi Steve, I am not sure I completely understand you UI suggestion, do you
> want it to show numerical values when you mouse in slicer view and the
> checkbox when your mouse is in the panel? It may be confusing. I can move
> the checkbox to the right of the numerical values to save vertical space if
> this sounds ok to you.
>
> Also I cannot reproduce the crash you reported on windows. Are you using
> Mac? I will try to add more checks in the code and perhaps you can try
> again.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Slicer/Slicer/pull/276#issuecomment-101432128>.
>

lassoan commented:
Do not put any settings in the data probe window itself! You can put the enable/disable checkbox in the Data Probe module (either above or below the slice view annotation settings).

ayarmarkovich commented:
Hi Andras. Would be nice to know the reason for marching orders :)

lassoan commented:
Lack of space due to data probe size is a constant issue. The situation should not be made worse if there is any alternative solution - and there are plenty.

The Data Probe module GUI is a nice option because people would likely to check out the Data Probe module for configuring the Data Probe. The application settings would be a good place, too (as a user I would look for the setting in the Application Settings or in the View menu).

ayarmarkovich commented:
Andras, Data Probe module has no relation to Data Probe panel and is a misnomer IMHO. The solution Steve proposed will not increase the size of the Data Probe panel.

lassoan commented:
Showing the zoomed image in place of the Slicer logo does not make the lack of space problem worse, so it would be OK for me.

If this option will be implemented then the zoomed display probably only needed to be disabled for performance reasons (when we display a live image in the slice viewer), and so it may be enough to allow disabling the feature programmatically (GUI might not be needed).

I agree that currently the Data Probe module is misused, as slice view settings belong to application settings. Data Probe module can be still used for configuring the Data Probe, if needed.

ayarmarkovich commented:
I changed the behavior as Steve suggested, so that zoomed view and checkbox do not  take any extra space in the Data Probe panel by default. Also added some bounds checks to try to fix crash on Mac.
Steve, see if you like this behavior and if you can still crash it (it never crashed on Windows, so I am not sure what was the problem).

pieper commented:
Hi Alex -

I like the new behavior but I am still seeing regular crashes and error
messages like this:

UpdateInformation invoked during another request.  Returning failure to
algorithm vtkExtractVOI(0x7ff491138880).


Maybe if you write some tests we can narrow it down?  Here's an example of
a test that emulates mouse movement:

https://github.com/pieper/CompareVolumes/blob/master/CompareVolumes.py#L985-L1033


-Steve


On Fri, May 15, 2015 at 11:07 AM, ayarmarkovich <notifications@github.com>
wrote:

> I changed the behavior as Steve suggested, so that zoomed view and
> checkbox do not take any extra space in the Data Probe panel by default.
> Also added some bounds checks to try to fix crash on Mac.
> Steve, see if you like this behavior and if you can still crash it (it
> never crashed on Windows, so I am not sure what was the problem).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Slicer/Slicer/pull/276#issuecomment-102424990>.
>

ayarmarkovich commented:
Steve, I was able to reproduce the crash on windows; it sometimes happened when you moved the mouse during loading a volume. Was that when you saw the crash? Adding more checks for ROI indices seemed to fix it. Could you try again with the latest version?

pieper commented:
Hi Alex -

It still crashes for me.  I didn't do anything but turn it on after the
volume was already loaded.

How about this: you can make a patched version for some of the folks in
Raul's group who want to use the feature and then when it's been stable for
a while we can put it in the trunk.  You'll just need to have them replace
the DataProbe.py script.

-Steve

[image: Inline image 1]

On Wed, May 20, 2015 at 7:59 PM, ayarmarkovich <notifications@github.com>
wrote:

> Steve, I was able to reproduce the crash on windows; it sometimes happened
> when you moved the mouse during loading a volume. Was that when you saw the
> crash? Adding more checks for ROI indices seemed to fix it. Could you try
> again with the latest version?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Slicer/Slicer/pull/276#issuecomment-104075241>.
>

ayarmarkovich commented:
Raul and Jinho have tested DataProbe on Mac and could not reproduce crashes reported by Steve.
Raul requested to add a cross-hair which I did. It is colored by the layout color. At this point I suggest merging into slicer to allow other people to test this functionality.

pieper commented:
Hi Alex -

Unless you have done something more to prevent the crash I reported I don't
think it should be merged into the trunk without more testing.

Did Raul and Jinho use release or debug builds?  I used a debug build and
that is what revealed the std::vector exception I reported earlier.  That
exception may be optimized out of a release build but I think it's
important to investigate.

-Steve

On Mon, May 25, 2015 at 12:22 PM, ayarmarkovich <notifications@github.com>
wrote:

> Raul and Jinho have tested DataProbe on Mac and could not reproduce
> crashes reported by Steve.
> Raul requested to add a cross-hair which I did. It is colored by the
> layout color. At this point I suggest merging into slicer to allow other
> people to test this functionality.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Slicer/Slicer/pull/276#issuecomment-105261451>.
>

ayarmarkovich commented:
Hi Steve,
I fixed the std::vector exception in one of the earlier commits. Perhaps you have not updated to it. 

Raul, Jinho, and I have tested in both debug and release mode on Mac and Windows and could not reproduce the crash. That makes me think that either you don't use the latest code or it is something specific to your environment.

Alex

pieper commented:
Hi Alex -

Which commit fixes the std::vector exception?

-Steve

On Wed, May 27, 2015 at 8:56 PM, ayarmarkovich <notifications@github.com>
wrote:

> Hi Steve,
> I fixed the std::vector exception in one of the earlier commits. Perhaps
> you have not updated to it.
>
> Raul, Jinho, and I have tested in both debug and release mode on Mac and
> Windows and could not reproduce the crash. That makes me think that either
> you don't use the latest code or it is something specific to your
> environment.
>
> Alex
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Slicer/Slicer/pull/276#issuecomment-106127722>.
>

ayarmarkovich commented:
Hi Steve,
This fixed the exception:
https://github.com/ayarmarkovich/Slicer/commit/6a23d5f0cf9f4bf24180f16fb8786c665c39bf08
The latest version should be good too.

pieper commented:
Okay, thanks.  If you are convinced it works fine then +1 to go ahead and
commit to the trunk.

-Steve

On Thu, May 28, 2015 at 9:05 PM, ayarmarkovich <notifications@github.com>
wrote:

> Hi Steve,
> This fixed the exception:
> ayarmarkovich@6a23d5f
> <https://github.com/ayarmarkovich/Slicer/commit/6a23d5f0cf9f4bf24180f16fb8786c665c39bf08>
> The latest version should be good too.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Slicer/Slicer/pull/276#issuecomment-106645931>.
>

ayarmarkovich commented:
committed to the trunk.

jcfr commented:
Thanks. Integrated in [r24284](http://viewvc.slicer.org/viewvc.cgi/Slicer4?view=revision&revision=24284)
